### PR TITLE
CORE-165 Add More Debugging Messages to Invalid YARD objects

### PR DESCRIPTION
**Pull request by cademarkegard**
- Added a logger for SwaggerYard

Todo:
- [x] Add specs to show when logging messages will be raised
- [x] Set logging level to WARN instead of DEBUG, allow manual override like YARD's

### CODE REVIEW

File: lib/swagger_yard/logging.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+module SwaggerYard
+  class Logger < ::Logger
+    def self.instance(pipe = STDOUT)
+      @logger ||= new(pipe)
+    end
+
+    def initialize(pipe, *args)
+      super(pipe, *args)
+      self.level = DEBUG
```
Comment (cademarkegard) [inline lib/swagger_yard/logging.rb]: set to WARN instead

File: lib/swagger_yard/api_declaration.rb [hunk 2]
```diff
@@ -25,6 +25,8 @@ def add_yard_object(yard_object)
           yard_object.children.each do |child_object|
             add_yard_object(child_object)
           end
+        else
+          swaggeryard_log.warn("Invalid controller object in file `#{yard_object.file}` near line #{yard_object.line}")
```
Comment (cademarkegard) [inline lib/swagger_yard/api_declaration.rb]: - This gets called a lot

File: lib/swagger_yard/global.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+require_relative 'logging'
+
+def swaggeryard_log
```
Comment (twrodriguez) [inline lib/swagger_yard/global.rb]: No globals plz

Comment (cademarkegard) [inline lib/swagger_yard/global.rb]: mmk!

File: lib/swagger_yard/api_declaration.rb [hunk 4]
```diff
@@ -36,37 +38,40 @@ def add_info(yard_object)
...
+      if tag = yard_object.tags.detect { |t| t.tag_name == 'resource_path' }
+        log.warn 'DEPRECATED: @resource_path tag is obsolete.'
       end
 
       # we only have api_key auth, the value for now is always empty array
-      @authorizations = Hash[yard_object.tags.
-                             select {|t| t.tag_name == "authorize_with"}.
-                             map(&:text).uniq.
-                             map {|k| [k, []]}]
+      @authorizations = Hash[yard_object.tags
```
Comment (twrodriguez) [inline lib/swagger_yard/api_declaration.rb]: replace with `each_with_object`

File: lib/swagger_yard/api_declaration.rb [hunk 5]
```diff
@@ -36,37 +38,40 @@ def add_info(yard_object)
...
+        return
+      end
 
       api = (apis[path] ||= Api.from_yard_object(yard_object, self))
       api.add_operation(yard_object)
     end
 
     def apis_hash
-      Hash[apis.map {|path, api| [path, api.operations_hash]}]
+      Hash[apis.map { |path, api| [path, api.operations_hash] }]
```
Comment (twrodriguez) [inline lib/swagger_yard/api_declaration.rb]: replace with `each_with_object`

File: lib/swagger_yard/model.rb [hunk 6]
```diff
@@ -47,25 +54,25 @@ def parse_tags(tags)
...
         }
       end
     end
 
     def to_h
       h = {
-        "type" => "object",
-        "properties" => Hash[@properties.map {|p| [p.name, p.to_h]}]
+        'type' => 'object',
+        'properties' => Hash[@properties.map { |p| [p.name, p.to_h] }]
```
Comment (twrodriguez) [inline lib/swagger_yard/model.rb]: replace with `each_with_object`

File: lib/swagger_yard/parameter.rb [hunk 7]
```diff
@@ -46,17 +46,16 @@ def from_path?
...
 
     def to_h
-      { "name"        => name,
-        "description" => description,
-        "required"    => required,
-        "in"          => param_type
-      }.tap do |h|
-        if h["in"] == "body"
-          h["schema"] = @type.to_h
+      { 'name'        => name,
```
Comment (twrodriguez) [inline lib/swagger_yard/parameter.rb]: This method is a mess. Can we add some variables or something?

File: lib/swagger_yard/resource_listing.rb [hunk 8]
```diff
@@ -39,19 +39,21 @@ def tag_objects
     end
 
     def model_objects
-      Hash[models.map {|m| [m.id, m.to_h]}]
+      Hash[models.map { |m| [m.id, m.to_h] }]
```
Comment (twrodriguez) [inline lib/swagger_yard/resource_listing.rb]: replace with `each_with_object`

File: lib/swagger_yard/resource_listing.rb [hunk 9]
```diff
@@ -39,19 +39,21 @@ def tag_objects
...
     end
 
     def security_objects
-      Hash[authorizations.map {|auth| [auth.name, auth.to_h]}]
+      Hash[authorizations.map { |auth| [auth.name, auth.to_h] }]
```
Comment (twrodriguez) [inline lib/swagger_yard/resource_listing.rb]: replace with `each_with_object`

File: lib/swagger_yard/type.rb [hunk 10]
```diff
@@ -44,42 +44,43 @@ def model_name
...
     end
 
     def to_h
       type = if ref?
-        { "$ref" => "#/definitions/#{name}"}
-      elsif enum?
-        { "type" => "string", "enum" => @enum }
-      else
-        json_type
+               { '$ref' => "#/definitions/#{name}" }
```
Comment (twrodriguez) [inline lib/swagger_yard/type.rb]: Can we clean this entire method up?

File: lib/swagger_yard/type.rb [hunk 11]
```diff
@@ -66,15 +66,17 @@ def json_type
...
     end
 
-    def to_h
-      type = if ref?
-               { '$ref' => "#/definitions/#{name}" }
-             elsif enum?
-               { 'type' => 'string', 'enum' => @enum }
-             else
-               json_type
+    def type
```
Comment (twrodriguez) [inline lib/swagger_yard/type.rb]: Should probably name this `swagger_type` instead. `type` seems to be used frequently in this class 😆

Comment (cademarkegard) [inline lib/swagger_yard/type.rb]: 👍 haha truth.

**Comment by twrodriguez** [comment]
LGTM 👍

**Pull request merged**